### PR TITLE
Both directions syncronization

### LINKED ISSUE #43: Bidirectional storage syncronization
Now only locally generated files are copied into underlying storage without checking for updates, which is bad in a numerous amount of cases. 

The suggestion is to perform synchronization in both directions and synchronize the only often changing(the specs) and newly created files.

### PR DESCRIPTION
**Pull request by Sammers21**
Storage synchronization has been improved: now changes are downloaded from the underlying storage before performing an update. 

Closes #43

### CODE REVIEW

**Comment by g4s8** [comment]
> Storage synchronization has been improved, now changes are downloaded from the underlying storage before performing an update

What was wrong with previos version of storage? I suppose we need a bug with details about it

**Comment by Sammers21** [comment]
> > Storage synchronization has been improved, now changes are downloaded from the underlying storage before performing an update
> 
> What was wrong with previos version of storage? I suppose we need a bug with details about it

@g4s8, before we've been copying local files to the storage without checking for an update, it means that multi-instance mode would be impossible without the change.

**Comment by 0crat** [comment]
This pull request [#41](https://github.com/artipie/gem-adapter/issues/41) is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/1853cb7b-278d-45bf-83b4-842e674e7b55); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by g4s8** [comment]
@0crat assign @victornoel

**Comment by g4s8** [comment]
@Sammers21 we need a ticket with a bug description to link this pull request, it's not clear why we can't push gems to multi instance artipie right now

**Comment by Sammers21** [comment]
@g4s8, I have already explained it.

**Comment by 0crat** [comment]
> @0crat assign @victornoel  ([here](https://github.com/artipie/gem-adapter/issues/41#issuecomment-628581046))

@g4s8 @g4s8[/z](https://www.0crat.com/u/g4s8) not enough funds available in the project, can't set budget of job [#41](https://github.com/artipie/gem-adapter/issues/41), see [§21](http://www.zerocracy.com/policy.html#21); @victornoel[/z](https://www.0crat.com/u/victornoel) will get no money on completion; in order to fix that, add funds to the project and assign the job again

**Comment by 0crat** [comment]
@0crat[/z](https://www.0crat.com/u/0crat) I see that you unassigned this issue; the order is still assigned to @victornoel[/z](https://www.0crat.com/u/victornoel) though; to cancel the order use `refuse`, as in [§6](http://www.zerocracy.com/policy.html#6)

**Comment by g4s8** [comment]
@Sammers21 I don't understand why we need this pull request

**Comment by Sammers21** [comment]
@g4s8, can you read the pull request description and my response?

**Comment by 0crat** [comment]
Job [#41](https://github.com/artipie/gem-adapter/issues/41) is already in scope

**Comment by 0crat** [comment]
Job [#41](https://github.com/artipie/gem-adapter/issues/41) is not in the agenda of @victornoel[/z](https://www.0crat.com/u/victornoel), can't set title

**Comment by 0crat** [comment]
GitHub issue is still open, won't close

File: src/main/ruby/AstoIndexer.rb [hunk 1]
```diff
@@ -383,7 +383,6 @@ def update_index
 
     if updated_gems.empty?
       say 'No new gems'
-      terminate_interaction 0
```
Comment (g4s8) [inline src/main/ruby/AstoIndexer.rb:386]: @Sammers21 why did you remove this line?

File: src/main/ruby/SubmitGem.rb [hunk 2]
```diff
@@ -38,40 +39,51 @@ def initialize(storage, fs)
...
+                  Key::From.new("prerelease_specs.4.8.gz"),
+                  Key::From.new("specs.4.8"),
+                  Key::From.new("specs.4.8.gz")
+              ]
+          )
+      )
+      Flowable::from_iterable(keys)
+    }
+    # Non-specs files, copy if not exists
+    diff = from.list(Key::From.new("quick"))
```
Comment (g4s8) [inline src/main/ruby/SubmitGem.rb:75]: @Sammers21 what is `"quick"`? please add some comments about "quick" if it has some special meaning

File: src/main/ruby/SubmitGem.rb [hunk 3]
```diff
@@ -38,40 +39,51 @@ def initialize(storage, fs)
...
+               .zipWith(
+                   to.list(Key::From.new("quick")),
+                   -> src, dest {
+                     src.removeAll(dest)
+                     src
+                   }
+               )
+               .flatMapPublisher { |keys| Flowable::from_iterable(keys) }
+    specs.mergeWith(diff).to_list.flatMapCompletable do |keys|
+      @@log.debug("Copy {}", keys)
```
Comment (g4s8) [inline src/main/ruby/SubmitGem.rb:85]: @Sammers21 is it a typo? why double `@@` here?

File: src/main/ruby/SubmitGem.rb [hunk 4]
```diff
@@ -38,40 +39,51 @@ def initialize(storage, fs)
...
+                  Key::From.new("latest_specs.4.8"),
+                  Key::From.new("latest_specs.4.8.gz"),
+                  Key::From.new("prerelease_specs.4.8"),
+                  Key::From.new("prerelease_specs.4.8.gz"),
+                  Key::From.new("specs.4.8"),
+                  Key::From.new("specs.4.8.gz")
+              ]
+          )
+      )
+      Flowable::from_iterable(keys)
```
Comment (g4s8) [inline src/main/ruby/SubmitGem.rb:72]: @Sammers21 it's not a good statement, most probably you need to return it?

File: src/main/ruby/SubmitGem.rb [hunk 5]
```diff
@@ -38,40 +39,51 @@ def initialize(storage, fs)
...
-                        Key::From.new("latest_specs.4.8.gz"),
-                        Key::From.new("prerelease_specs.4.8"),
-                        Key::From.new("prerelease_specs.4.8.gz"),
-                        Key::From.new("specs.4.8"),
-                        Key::From.new("specs.4.8.gz")
-                    ]
-                )
-            )
-        ).to_list
+  def sync(from, to)
```
Comment (g4s8) [inline src/main/ruby/SubmitGem.rb:57]: @Sammers21 what is the types of `from` and `to`? please add these details to comments

File: src/main/ruby/SubmitGem.rb [hunk 6]
```diff
@@ -38,40 +39,51 @@ def initialize(storage, fs)
...
+      )
+      Flowable::from_iterable(keys)
+    }
+    # Non-specs files, copy if not exists
+    diff = from.list(Key::From.new("quick"))
+               .zipWith(
+                   to.list(Key::From.new("quick")),
+                   -> src, dest {
+                     src.removeAll(dest)
+                     src
```
Comment (g4s8) [inline src/main/ruby/SubmitGem.rb:80]: @Sammers21 why `src` here?

File: src/main/ruby/AstoIndexer.rb (hunk 1 above)
Comment (Sammers21) [reply src/main/ruby/AstoIndexer.rb:386]: Program should not terminate here

File: src/main/ruby/SubmitGem.rb (hunk 2 above)
Comment (Sammers21) [reply src/main/ruby/SubmitGem.rb:75]: No special meaning, it is a standard directory name for a gem repository

File: src/main/ruby/SubmitGem.rb (hunk 3 above)
Comment (Sammers21) [reply src/main/ruby/SubmitGem.rb:85]: @@ refers to a static variable

File: src/main/ruby/SubmitGem.rb (hunk 4 above)
Comment (Sammers21) [reply src/main/ruby/SubmitGem.rb:72]: Last expression is always returned in ruby

File: src/main/ruby/SubmitGem.rb (hunk 6 above)
Comment (Sammers21) [reply src/main/ruby/SubmitGem.rb:80]: The same as `return src`

File: src/main/ruby/SubmitGem.rb (hunk 5 above)
Comment (Sammers21) [reply src/main/ruby/SubmitGem.rb:59]: Done

**Comment by Sammers21** [comment]
@g4s8, review the PR once again please

File: src/main/ruby/SubmitGem.rb (hunk 2 above)
Comment (g4s8) [reply src/main/ruby/SubmitGem.rb:78]: @Sammers21 it's not clear from the name (why quick is a repository?) so I'd add a comment about it

Comment (Sammers21) [reply src/main/ruby/SubmitGem.rb:78]: @g4s8 , quick is not a repository. Read it: https://blog.packagecloud.io/eng/2015/12/15/rubygem-index-internals/

File: pom.xml [hunk 7]
```diff
@@ -79,7 +79,7 @@ SOFTWARE.
     <dependency>
       <groupId>com.artipie</groupId>
       <artifactId>asto</artifactId>
-      <version>0.19.1</version>
+      <version>1.0-SNAPSHOT</version>
```
Comment (g4s8) [inline pom.xml:82]: @Sammers21 why snapshot here?

**Comment by 0crat** [comment]
Job [#41](https://github.com/artipie/gem-adapter/issues/41) is not in the agenda of @victornoel[/z](https://www.0crat.com/u/victornoel), can't set title

**Comment by 0crat** [comment]
Job was finished in 3 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Job [#41](https://github.com/artipie/gem-adapter/issues/41) is not in the agenda of @victornoel[/z](https://www.0crat.com/u/victornoel), can't set title

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@g4s8 @Sammers21 please make sure you start all your comments with the name of the user they are referred to, see: [1](https://github.com/artipie/gem-adapter/pull/41#issuecomment-627935433); [2](https://github.com/artipie/gem-adapter/pull/41#discussion_r425141498)

**Comment by sereshqua** [comment]
@0crat quality acceptable
